File: android/app/src/main/kotlin/com/stillmoment/domain/models/TimerDisplayState.kt
package com.stillmoment.domain.models

import java.util.Locale

/**
 * Complete UI state for the timer display.
 *
 * This data class contains all data needed to render the timer UI.
 * All computed properties are pure functions of the stored properties,
 * making this state fully testable without mocks.
 */
data class TimerDisplayState(
    /** Current timer state */
    val timerState: TimerState = TimerState.Idle,
    /** Selected duration in minutes (1-60) */
    val selectedMinutes: Int = MeditationSettings.DEFAULT_DURATION_MINUTES,
    /** Remaining time in seconds */
    val remainingSeconds: Int = 0,
    /** Total duration in seconds */
    val totalSeconds: Int = 0,
    /** Remaining preparation seconds (15, 14, 13... 0) */
    val remainingPreparationSeconds: Int = 0,
    /** Progress value (0.0 - 1.0) */
    val progress: Float = 0f,
    /** Current affirmation index (rotates between sessions) */
    val currentAffirmationIndex: Int = 0,
    /** Whether an interval gong was already played for current interval */
    val intervalGongPlayedForCurrentInterval: Boolean = false
) {
    // MARK: - Computed Properties (Pure, Testable)

    /** Whether currently in preparation phase */
    val isPreparation: Boolean
        get() = timerState == TimerState.Preparation

    /** Returns true if timer can be started */
    val canStart: Boolean
        get() = timerState == TimerState.Idle && selectedMinutes > 0

    /** Returns true if timer can be paused */
    val canPause: Boolean
        get() = timerState == TimerState.Running

    /** Returns true if timer can be resumed */
    val canResume: Boolean
        get() = timerState == TimerState.Paused

    /** Returns true if timer can be reset */
    val canReset: Boolean
        get() = timerState != TimerState.Idle

    /** Formatted time string (MM:SS or preparation seconds) */
    val formattedTime: String
        get() =
            if (isPreparation) {
                "$remainingPreparationSeconds"
            } else {
                val minutes = remainingSeconds / 60
                val seconds = remainingSeconds % 60
                String.format(Locale.ROOT, "%02d:%02d", minutes, seconds)
            }

    companion object {
        /** Initial state for a fresh timer */
        val Initial = TimerDisplayState()

        /**
         * Creates a state with custom selected minutes (for loading from settings).
         */
        fun withDuration(minutes: Int): TimerDisplayState {
            return Initial.copy(
                selectedMinutes = MeditationSettings.validateDuration(minutes)
            )
        }
    }
}

File: ios/StillMoment/Domain/Models/TimerDisplayState.swift
//
//  TimerDisplayState.swift
//  Still Moment
//
//  Domain Model - Timer Display State for Reducer Pattern
//

import Foundation

/// Complete UI state for the timer display
///
/// This struct contains all data needed to render the timer UI.
/// All computed properties are pure functions of the stored properties,
/// making this state fully testable without mocks.
struct TimerDisplayState: Equatable {
    // MARK: - Stored Properties

    /// Current timer state
    var timerState: TimerState

    /// Selected duration in minutes (1-60)
    var selectedMinutes: Int

    /// Remaining time in seconds
    var remainingSeconds: Int

    /// Total duration in seconds
    var totalSeconds: Int

    /// Remaining preparation seconds (e.g., 15, 14, 13... 0)
    var remainingPreparationSeconds: Int

    /// Progress value (0.0 - 1.0)
    var progress: Double

    /// Current affirmation index (rotates between sessions)
    var currentAffirmationIndex: Int

    /// Whether an interval gong was already played for current interval
    var intervalGongPlayedForCurrentInterval: Bool

    // MARK: - Computed Properties (Pure, Testable)

    /// Whether currently in preparation phase
    var isPreparation: Bool {
        self.timerState == .preparation
    }

    /// Returns true if timer can be started
    var canStart: Bool {
        self.timerState == .idle && self.selectedMinutes > 0
    }

    /// Whether the timer is actively running (for UI display)
    /// Returns true during start gong, introduction, silent meditation and endGong phases
    /// (no visual difference between these phases per design — ring full, 00:00 during endGong)
    var isRunning: Bool {
        self.timerState == .running || self.timerState == .introduction
            || self.timerState == .startGong || self.timerState == .endGong
    }

    /// Formatted time string (MM:SS or preparation seconds)
    var formattedTime: String {
        if self.isPreparation {
            return "\(self.remainingPreparationSeconds)"
        }
        let minutes = self.remainingSeconds / 60
        let seconds = self.remainingSeconds % 60
        return String(format: "%02d:%02d", minutes, seconds)
    }

    // MARK: - Static Factory

    /// Initial state for a fresh timer
    static let initial = TimerDisplayState(
        timerState: .idle,
        selectedMinutes: 10,
        remainingSeconds: 0,
        totalSeconds: 0,
        remainingPreparationSeconds: 0,
        progress: 0.0,
        currentAffirmationIndex: 0,
        intervalGongPlayedForCurrentInterval: false
    )

    /// Creates a state with custom selected minutes (for loading from settings)
    static func withDuration(minutes: Int, introductionId: String? = nil) -> TimerDisplayState {
        var state = self.initial
        state.selectedMinutes = MeditationSettings.validateDuration(minutes, introductionId: introductionId)
        return state
    }
}
